File: lua/noise.lua
local ffi = require("ffi")
local lib = require("micro/lua/libmicro")

ffi.cdef([[
    typedef struct NoiseCfg
    {
      int width;
      int height;
      float frequency;
      float amplitude;
      float offsetX;
      float offsetY;
      int octaves;
      float persistence;
      float lacunarity;
      bool usePeriodic;
      int periodX;
      int periodY;
      bool normalize;
    } NoiseCfg;

    void microNoisePelin2(float *data, const NoiseCfg cfg);
]])

local Noise = {}

---This function generates a 2D noise map
---using the Perlin noise algorithm.
---@param width number
---@param height number
---@param frequency number
---@param amplitude number
---@param offsetX number
---@param offsetY number
---@param octaves number
---@param persistence number
---@param lacunarity number
---@param usePeriodic boolean
---@param periodX number
---@param periodY number
---@param normalize boolean
function Noise.generate(
	width,
	height,
	frequency,
	amplitude,
	offsetX,
	offsetY,
	octaves,
	persistence,
	lacunarity,
	usePeriodic,
	periodX,
	periodY,
	normalize
)
	local cfg = ffi.new("NoiseCfg")
	cfg.width = width
	cfg.height = height
	cfg.frequency = frequency
	cfg.amplitude = amplitude
	cfg.offsetX = offsetX
	cfg.offsetY = offsetY
	cfg.octaves = octaves
	cfg.persistence = persistence
	cfg.lacunarity = lacunarity
	cfg.usePeriodic = usePeriodic
	cfg.periodX = periodX
	cfg.periodY = periodY
	cfg.normalize = normalize

	local dataPtr = ffi.new("float[?]", width * height)

	lib.microNoisePelin2(dataPtr, cfg)

	return dataPtr
end

return Noise

File: src/micro/core/Noise.c
#include "Noise.h"
#include "../util/debug.h"
#include "../util/perlin_noise.h"
#include <math.h>

static float fractalNoise2(float x, float y, const NoiseCfg *cfg)
{
  float total = 0.0f;
  float frequency = cfg->frequency;
  float amplitude = cfg->amplitude;
  float maxAmplitude = 0.0f;

  for (int i = 0; i < cfg->octaves; ++i)
  {
    float nx = x * frequency;
    float ny = y * frequency;

    float noise = cfg->usePeriodic ? pnoise2(nx, ny, cfg->periodX, cfg->periodY)
                                   : noise2(nx, ny);

    total += noise * amplitude;
    maxAmplitude += amplitude;

    amplitude *= cfg->persistence;
    frequency *= cfg->lacunarity;
  }

  return total / maxAmplitude;
}

void microNoisePelin2(float *data, const NoiseCfg cfg)
{
  assert(cfg.width > 0);
  assert(cfg.height > 0);
  assert(cfg.frequency != 0.0f);
  assert(cfg.amplitude != 0.0f);
  assert(cfg.octaves >= 1);

  for (int y = 0; y < cfg.height; ++y)
  {
    for (int x = 0; x < cfg.width; ++x)
    {
      float fx = (x + cfg.offsetX) / (float)cfg.width;
      float fy = (y + cfg.offsetY) / (float)cfg.height;

      float noise = fractalNoise2(fx, fy, &cfg);

      if (cfg.normalize)
        noise = (noise + 1.0f) * 0.5f;

      data[x + y * cfg.width] = noise;
    }
  }
}
